### ISSUE TITLE
The approach is error-prone

### ISSUE THREAD

**Issue by UnkindPartition** [ocharles/exhaustive#2]
I agree with the premise of this library: let's make errors harder to commit.

Unfortunately, this specific approach makes it even easier to commit certain types of errors. In your example, if I swap `ETrue` and `EFalse` around, not only will the compiler not catch it, but it also becomes harder to spot the error when reading the code than it would be in the original, naive code:

```
  ETrue <$ symbol "False"
, EFalse <$ symbol "True"
```

vs

```
  construct (\f -> f <$ symbol "False") :*
  construct (\f -> f <$ symbol "True") :*
```

**Comment by ocharles** [comment]
Yes, it is not perfect - but what do you suggest - drop the library
entirely? :) I'm aware of this, and its a caveat of the approach. It could
do with more prominent documentation though.

**Comment by UnkindPartition** [comment]
I don't have any suggestions so far; just wanted to point this out and start a discussion.

**Comment by ocharles** [comment]
Ok, that's fair enough. One thing I would have liked to be able to do was:

``` haskell
constructWith ETrue (\eTrue -> eTrue <$ symbol "True")
```

And then have some sort of ordering information so I can at least require things are written in the same order as the data type. However, being given `ETrue` is no good - it's just a function by the time it reaches me, so I have no way to reflect anything useful out of it.

I will keep thinking about this though, maybe there is a solution.

**Comment by UnkindPartition** [comment]
Yes, that's similar to what I had in mind.

You could use TH to generate constructor-like tokens based on constructor names. The types of these tokens could involve Peano indexes, so that you could validate this at compile-time.

**Comment by ocharles** [comment]
I suppose I could generate my own symbols, that's a good point. I feel like falling back to TH is the last resort - as we've already required the user to derive a `Generic` instance - I'd rather not have them jump through more hoops. Unfortunately, there might not be any other choices.

**Comment by ocharles** [comment]
I wonder if we could quasiquote the original constructor names...

**Comment by UnkindPartition** [comment]
We definitely could; I'm not sure how much it'd gain us, however.

**Comment by ocharles** [comment]
Well, I'm wondering if that would allow us to learn more about the symbol in question, and then we could have the quasiquoter produce something like `Proxy :: Proxy (S (S Z))`

**Comment by UnkindPartition** [comment]
Yes, I see what you mean (and it _is_ possible); I'm just saying that it's not clear how much an improvement over symbol-generation it is.

**Comment by ocharles** [comment]
Great! With TH and custom symbols there is one obvious problem - what do you call them? You can't reuse the existing name, so you have to pick a different name. However, by doing so, using the library becomes more annoying as you can't just read out the original constructors. Also, you have to enable Template Haskell (which has ordering implications in the file) and write an extra line before you can do anything. It seems quasiquoting could remove all of those problems.

**Comment by UnkindPartition** [comment]
By the way, by generating symbols or using quasiquotes we could simplify the parsing code. Here's a very sketchy prototype:

``` haskell
data Nat = Z | S Nat
data P (n :: Nat) a
data Q (n :: Nat) a
instance Applicative (P n)
-- Alternative-like definitions
empty :: Q Z
infixl 3 <|>
(<|>) :: Q n a -> P n a -> Q (S n) a
extractParser :: (n ~ NumberOfConstructors a) => Q n a -> Parser a
```

The quasiquote or symbol generator would produce a `P n (Int -> String -> User)` for the nth data constructor of the corresponding type.

**Comment by UnkindPartition** [comment]
(I've omitted the boring details like parameterizing on the functor.)

**Comment by UnkindPartition** [comment]
With QuasiQuotes, you still have to enable an extension (a different one); and the QQ noise is scattered around the code, instead of being concentrated in one TH line. But I'm not opposed to it; just saying that it isn't a clear win.

**Comment by ocharles** [comment]
I had a play with this, and I think I'm barking up the wrong tree with quasiquoters - but splicing in a peano number with Template Haskell is looking somewhat promising. Here's what I have:

``` haskell
parseExpr :: Parser Expr
parseExpr =
  produceFirst $
  cons $(con 'ETrue) (construct (\f -> f <$ symbol "True")) $
  cons $(con 'EFalse) (construct (\f -> f <$ symbol "False")) $
  cons $(con 'EIf) (construct (\f -> f <$> (symbol "if" *> parseExpr)
                                      <*> (symbol "then" *> parseExpr)
                                      <*> (symbol "else" *> parseExpr))) $
  Nil
```

This type checks that you use all constructors in the same order as the original data definition. We have...

``` haskell
cons :: (Length code ~ (n + Length xs)) => Proxy n -> Producer f code x -> NP (Producer f code) xs -> NP (Producer f code) (x ': xs)
```

This means forces a total ordering on `cons`.

**Comment by ocharles** [comment]
I think I might run with this a bit more, such that the final syntax is something like:

``` haskell
parseExpr :: Parser Expr
parseExpr =
  produceFirst $
  construct $(con 'ETrue) (\f -> f <$ symbol "True") :&
  construct $(con 'EFalse) (\f -> f <$ symbol "False") :&
  construct $(con 'EIf) (\f -> f <$> (symbol "if" *> parseExpr)
                                 <*> (symbol "then" *> parseExpr)
                                 <*> (symbol "else" *> parseExpr)) :&
  Done
```

**Comment by UnkindPartition** [comment]
What I was getting at was that you should be able to get rid of `f`:

``` haskell
construct ($(con 'ETrue) <*> symbol "True") :&
```

**Comment by UnkindPartition** [comment]
Also, wouldn't a quasiquoter let you write something like

``` haskell
construct ([con|ETrue|] <*> symbol "True") :&
```

?

**Comment by ocharles** [comment]
Right, its a bit more work to make that happen but it might be doable. Will
carry on exploring.

**Comment by ocharles** [comment]
Alright, got further! I can now do the following:

``` haskell
parseExpr' :: Parser Expr
parseExpr' =
  produceFirst $
  ($(con 'ETrue) <$ symbol "True") &:
  ($(con 'EFalse) <$ symbol "False") &:
  ($(con 'EIf) <$> (symbol "if" *> parseExpr)
               <*> (symbol "if" *> parseExpr)
               <*> (symbol "if" *> parseExpr)) &:
  exhausted
```

Switching any of the elements passed to `produceFirst`, removing an element, or duplicating an element all result in a type error - as expected.

**Comment by ocharles** [comment]
At this point I might as well fully commit to TH, which means I can drop all the `generics-sop` stuff.

**Comment by ocharles** [comment]
OK, I do still need some generics stuff - I think https://ocharles.org.uk/tmp/exhaustive/Control-Exhaustive.html might be the best I can do for now. Do you have any thoughts on that, or shall I go ahead and release this?

**Comment by UnkindPartition** [comment]
You haven't commented on my indexed applicative approach described above. So I'm not sure if you didn't like it or didn't understand it :) I think it would be more straightforward.

**Comment by ocharles** [comment]
I understood it, but I can't see how it's workable. It either requires a
new version of the applicative combinators, or it requires users lift
everything into my applicative. Both are too intrusive imo.

**Comment by UnkindPartition** [comment]
Mmm, yeah, you're right.

And how do you feel about `[con|ETrue|]` instead of `$(con 'ETue)`?

**Comment by ocharles** [comment]
I'll have to see if I can actually get the information I want from that, though the nice thing about the TH is that it is a little more syntax aware (I already get predictable errors for typos, and refactoring tools know that I'm talking about a type). On the downside, it annoyingly forces an ordering of top level declarations.

I'll have a look and see if the quasiquoter works.

**Comment by ocharles** [comment]
As a note to myself... https://ghc.haskell.org/trac/ghc/ticket/9813?cversion=1&cnum_hist=5 is looking like a bastard of a bug though... it forces users of _this_ library to insert a random `$(return [])` top level splice, just to get things to compile. Urgh.

**Comment by UnkindPartition** [comment]
Yeah, that's ugly :/

**Comment by ocharles** [comment]
It looks like 9813 bites us even if we use a quasiquoter - in order to use `[con|ETrue|]` you have to use Template Haskell anyway to break out into a new declaration group and actually bring `ETrue` into scope. Due to this, I think I'll stick with using TH rather than a quasiquoter.

**Comment by ocharles** [comment]
Ok, I think I'm going to go with https://ocharles.org.uk/tmp/exhaustive/Control-Exhaustive.html -- any final feedback? I think this release would address the problem reported in this issue.

**Comment by UnkindPartition** [comment]
Sure, go ahead.

**Comment by ocharles** [comment]
New approach released as 1.1.0.
